utils/measure/measure/util/measure_util.py

```diff
@@ -21,21 +26,113 @@ def __init__(self, power_meter: PowerMeter, config: MeasureConfig) -> None:
+
+        This function calculates the average power or resistance by taking multiple readings over
+        the specified duration. If `measure_resistance` is True, the function computes resistance
+        using the formula R = V^2 / P, where V is the voltage and P is the power. If a dummy load
+        resistive value is set, the power consumption of the dummy load calculated for each
+        measurement, depending on the current voltage and is subtracted from the power
+        measurements.
+
+        Args:
+            duration (int): The time duration (in seconds) over which to take measurements.
+            measure_resistance (bool): Whether to measure resistance instead of power. Defaults to False.
+
+        Returns:
+            float: The average resistance (in Ω) or power (in W) over the measurement duration.
+
+        Raises:
+            UnsupportedFeatureError: If `measure_resistance` is True but the power meter does not
+                                      support voltage measurements.
+
+        Error handling:
+            - If voltage measurements are not supported, the program will terminate with an appropriate
+              error message, if measuring power consumption and a dummy load resistive value is set.
+            - For resistance measurements, voltage values below 1 are treated as invalid, and the program
+              exits to avoid incorrect calculations.
+            - Ignores single measurements of <= 0 W.
+        """
+        _LOGGER.info("Measuring average %s over %s seconds", "resistance" if measure_resistance else "power", duration)
         start_time = time.time()
         readings: list[float] = []
+
+        first_measurement = True
+
         while (time.time() - start_time) < duration:
+            if first_measurement:
+                first_measurement = False
+            else:
+                # sleep time exceeds duration
+                if not ((time.time() - start_time + self.config.sleep_time) < duration):
+                    break
+                time.sleep(self.config.sleep_time)
+
+            if measure_resistance:
+                result = self.power_meter.get_power(include_voltage=True)
+                power, voltage = result.power, result.voltage
+
+                if voltage < 1:
+                    _LOGGER.error("Error during measurement: Voltage measurement returned zero. Aborting measurement.")
+                    exit(1)
+
+                if round(power, 2) == 0:
+                    _LOGGER.warning("Invalid measurement: power: %.2f W, voltage: %.2f", power, voltage)
+                    continue
+
+                resistance = round((voltage**2) / power, 4)
+                readings.append(resistance)
+                _LOGGER.debug(
+                    "Measured resistance: %.2f Ω; measured power: %.2f W, voltage: %.2f",
+                    resistance,
+                    power,
+                    voltage,
+                )
+                continue
```

Comment:
We really need to bring the complexity of this loop down. Currently it's full of if/else, nesting, continue, break statements, which make it hard to get a mental model and difficult to debug.
Nice start would be to extract this code to seperate method `measure_resistance`, to keep the main while loop logic more concise and much easier to read. We can reduce the LOC of this method quite easily this way.
Also it allows for reusability if we need to measure resistance somewhere in the code, just call `measure_resistance`.

Reply:
Agreed. But it would make sense to cut down the two measurement functions take_average_measurement() and take_measurement() down to a single measurement function and restructure this code in the process. 

So take_average_measurement() calculates via the sleep interval, how many measurements will fit into the time period and then just asks take_measurement() for the number of measurements and returns the average back, like explained [above](https://github.com/bramstroker/homeassistant-powercalc/pull/2810#discussion_r1895934175).

This would remove the need for a lot of duplicate code here about error handling and we could handle the different measurements in subfunctions of take_measurement() or private methods.

Reply:
See my code from which I pasted below from chatGPT.
It has extracted into dedicated methods with single purpose out of the time loop.
So I methods are also reusable here now.
When still not reusable extract more methods and/or introduce extra param in the method to control the flow.